rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java

```diff
@@ -211,33 +211,38 @@ public MavenMetadata downloadMetadata(GroupArtifactVersion gav, @Nullable Resolv
             if (result == null) {
                 // Not in the cache, attempt to download it.
+                String scheme = URI.create(repo.getUri()).getScheme();
+                String uri = repo.getUri() + (repo.getUri().endsWith("/") ? "" : "/") +
+                        gav.getGroupId().replace('.', '/') + '/' +
+                        gav.getArtifactId() + '/' +
+                        (gav.getVersion() == null ? "" : gav.getVersion() + '/') +
+                        "maven-metadata.xml";
+
                 boolean cacheEmptyResult = false;
-                try {
-                    String scheme = URI.create(repo.getUri()).getScheme();
-                    String uri = repo.getUri() + (repo.getUri().endsWith("/") ? "" : "/") +
-                                 gav.getGroupId().replace('.', '/') + '/' +
-                                 gav.getArtifactId() + '/' +
-                                 (gav.getVersion() == null ? "" : gav.getVersion() + '/') +
-                                 "maven-metadata.xml";
-
-                    if ("file".equals(scheme)) {
-                        // A maven repository can be expressed as a URI with a file scheme
-                        Path path = Paths.get(URI.create(uri));
-                        if (Files.exists(path)) {
-                            result = Optional.of(MavenMetadata.parse(Files.readAllBytes(path)));
+                for (HttpSender.Request request : requestAsAuthenticatedAndAsAnonymous(repo, uri)) {
+                    cacheEmptyResult = false;
+                    try {
+                        if ("file".equals(scheme)) {
+                            // A maven repository can be expressed as a URI with a file scheme
+                            Path path = Paths.get(URI.create(uri));
+                            if (Files.exists(path)) {
+                                result = Optional.of(MavenMetadata.parse(Files.readAllBytes(path)));
+                            }
+                        } else {
+                            byte[] responseBody = sendRequest(request);
+                            result = Optional.of(MavenMetadata.parse(responseBody));
+
                         }
-                    } else {
-                        byte[] responseBody = requestAsAuthenticatedOrAnonymous(repo, uri);
-                        result = Optional.of(MavenMetadata.parse(responseBody));
-                    }
-                } catch (HttpSenderResponseException e) {
-                    repositoryResponses.put(repo, e.getMessage());
-                    if (e.isClientSideException()) {
-                        //If we have a 400-404, cache an empty result.
-                        cacheEmptyResult = true;
+                        break;
+                    } catch (HttpSenderResponseException e) {
+                        repositoryResponses.put(repo, e.getMessage());
```

Comment:
since `repositoryResponses` is a `Map`, and the `MavenRepository` is the same for both the authenticated and anonymous request, wouldn't the first failure response get overwritten in the `Map` at this line in the case where both requests fail?

Reply:
That's true, unfortunately. Maybe the PR is taking this too far and a better approach would be to change which exception is thrown from `MavenPomDownloader#requestAsAuthenticatedOrAnonymous`. Instead of throwing the 401 exception from the retry, throw the original one that caused the retry? After all it is kinda expected that the retry would fail with a 401.

Reply:
Hmm -- IIRC, some remote repo implementations return 404 for requests to artifacts which you aren't authorized to see, not 401. But I think something like that could still work, ie, deciding which exception is the "interesting" one based on the status codes from each

Reply:
I'll try to come up with something later. I'm busy with other stuff next week.